libyul/backends/evm/SSAControlFlowGraph.cpp

```diff
@@ -209,33 +245,40 @@ class SSACFGPrinter
 		}
 	}
 
-	void printFunction(SSACFG const& _cfg, size_t const _functionIndex, Scope::Function const& _fun)
+	void printFunction(Scope::Function const& _fun)
 	{
 		static auto constexpr returnsTransform = [](auto const& functionReturnValue) { return functionReturnValue.get().name.str(); };
 		static auto constexpr argsTransform = [](auto const& arg) { return fmt::format("v{}", std::get<1>(arg).value); };
-		m_result << "FunctionEntry_" << _fun.name.str() << "_" << _cfg.entry.value << " [label=\"";
-		if (!_cfg.returns.empty())
-			m_result << fmt::format("function {0}:\n {1} := {0}({2})", _fun.name.str(), fmt::join(_cfg.returns | ranges::views::transform(returnsTransform), ", "), fmt::join(_cfg.arguments | ranges::views::transform(argsTransform), ", "));
+		m_result << "FunctionEntry_" << _fun.name.str() << "_" << m_cfg.entry.value << " [label=\"";
+		if (!m_cfg.returns.empty())
+			m_result << fmt::format("function {0}:\n {1} := {0}({2})", _fun.name.str(), fmt::join(m_cfg.returns | ranges::views::transform(returnsTransform), ", "), fmt::join(m_cfg.arguments | ranges::views::transform(argsTransform), ", "));
 		else
-			m_result << fmt::format("function {0}:\n {0}({1})", _fun.name.str(), fmt::join(_cfg.arguments | ranges::views::transform(argsTransform), ", "));
+			m_result << fmt::format("function {0}:\n {0}({1})", _fun.name.str(), fmt::join(m_cfg.arguments | ranges::views::transform(argsTransform), ", "));
 		m_result << "\"];\n";
-		m_result << "FunctionEntry_" << _fun.name.str() << "_" << _cfg.entry.value << " -> Block" << _functionIndex << "_" << _cfg.entry.value << ";\n";
-		printBlock(_cfg, _cfg.entry, _functionIndex);
+		m_result << "FunctionEntry_" << _fun.name.str() << "_" << m_cfg.entry.value << " -> Block" << m_functionIndex << "_" << m_cfg.entry.value << ";\n";
+		printBlock(m_cfg.entry);
 	}
 
+	SSACFG const& m_cfg;
+	size_t m_functionIndex;
+	SSACFGLiveness const* m_liveness;
 	std::stringstream m_result{};
 };
 }
 
-std::string SSACFG::toDot(bool _includeDiGraphDefinition, std::optional<size_t> _functionIndex) const
+std::string SSACFG::toDot(
+	bool _includeDiGraphDefinition,
+	std::optional<size_t> _functionIndex,
+	SSACFGLiveness const* _liveness
+) const
 {
 	std::ostringstream output;
 	if (_includeDiGraphDefinition)
 		output << "digraph SSACFG {\nnodesep=0.7;\ngraph[fontname=\"DejaVu Sans\"]\nnode[shape=box,fontname=\"DejaVu Sans\"];\n\n";
 	if (function)
-		output << SSACFGPrinter(*this, _functionIndex ? *_functionIndex : static_cast<size_t>(1), *function);
+		output << SSACFGPrinter(*this, _functionIndex ? *_functionIndex : static_cast<size_t>(1), *function, _liveness);
```

Comment:
I just noticed this now, but why `1` if `_functionIndex` is null? In fact, why do we even need `functionIndex` in `SSACFGPrinter` at all? Doesn't the `SSACFG` already contain all the functions?

Reply:
SSACFG contains only one function (or the main graph), ControlFlow contains all of it :) the indices are needed to disambiguate the nodes in the dot graph